modules/sugar/src/test/ts/browser/ShadowDomTest.ts

```diff
@@ -48,3 +34,117 @@ shadowDomTest('ShadowDom - SelectorFind.descendant', () => {
     });
   }));
 });
+
+const shouldBeShadowRoot = (n: RootNode) => {
+  Assert.eq('should be shadow root', true, ShadowDom.isShadowRoot(n));
+  Assert.eq('should not be document', false, ShadowDom.isDocument(n));
+};
+
+const shouldBeDocument = (n: RootNode) => {
+  Assert.eq('should not be shadow root', false, ShadowDom.isShadowRoot(n));
+  Assert.eq('should be document', true, ShadowDom.isDocument(n));
+};
+
+UnitTest.test('getRootNode === document on normal element in dom', () => {
+  withNormalElement((div) => {
+    Assert.eq('should be document', Document.getDocument(), ShadowDom.getRootNode(div), tElement);
+  });
+});
+
+UnitTest.test('getRootNode(document) === document on normal element in dom', () => {
+  withNormalElement(() => {
+    Assert.eq('should be document', Document.getDocument(), ShadowDom.getRootNode(Document.getDocument()), tElement);
+  });
+});
+
+UnitTest.test('isDocument(getRootNode) === true on normal element in dom', () => {
+  withNormalElement((div) => {
+    shouldBeDocument(ShadowDom.getRootNode(div));
+  });
+});
+
+UnitTest.test('document is document', () => {
+  shouldBeDocument(Document.getDocument());
+});
+
+if (ShadowDom.isSupported()) {
+  UnitTest.test('getRootNode === shadowroot on element in shadow root', () => {
+    withShadowElement((sr, innerDiv) => {
+      Assert.eq('should be shadowroot', sr, ShadowDom.getRootNode(innerDiv), tElement);
+    });
+  });
+
+  UnitTest.test('getRootNode(shadowroot) === shadowroot', () => {
+    withShadowElement((sr) => {
+      Assert.eq('should be shadowroot', sr, sr, tElement);
+    });
+  });
+
+  UnitTest.test('shadow root is shadow root', () => {
+    withShadowElement((sr, innerDiv) => {
+      shouldBeShadowRoot(ShadowDom.getRootNode(innerDiv));
+      shouldBeShadowRoot(sr);
+    });
+  });
+}
+
+UnitTest.test('getRootNode in iframe', () => {
+  withIframe((div, iframe, cw) => {
+    Assert.eq('should be inner doc', cw.document, ShadowDom.getRootNode(div).dom(), Testable.tStrict);
+  });
+});
+
+UnitTest.test('isDocument in iframe', () => {
+  withIframe((div, iframe, cw) => {
+    shouldBeDocument(Element.fromDom(cw.document));
+  });
+});
+
+// TODO: this should be in sand
```

Comment:
We specifically don't want this in sand, as it's not useful in the wild and would cause code bloat... if anything it should be in Agar (or similar).

Reply:
Fair enough. We really need to address the sand bloat.

Reply:
Fixed.